src/premiumvideodevice/PremiumVideoTransformDevice.ts

```diff
@@ -0,0 +1,91 @@
+import Device from '../devicecontroller/Device';
+import VideoTransformDevice from '../devicecontroller/VideoTransformDevice';
+import Logger from '../logger/Logger';
+import PremiumVideoStreamHandler from './PremiumVideoStreamHandler';
+import PremiumVideoEffectDriver from './PremiumVideoEffectDriver';
+
+/**
+ * [[PremiumVideoTransformDevice]] will allow us to transform a regular device
+ * into a format that will allow us to intercept it's MediaStream via a
+ * [[PremiumVideoStreamHandler]] and apply frame-level effects via a
+ * [[PremiumVideoEffectDriver]]
+ */
+export default class PremiumVideoTransformDevice implements VideoTransformDevice {
+  private inputMediaStream: MediaStream;
+  private premiumVideoStreamHandler: PremiumVideoStreamHandler;
+
+  constructor(
+    private logger: Logger,
+    private device: Device,
+    private videoEffectDriver: PremiumVideoEffectDriver
+  ) {
+    // initialize the stream handler
+    this.premiumVideoStreamHandler = new PremiumVideoStreamHandler(logger, videoEffectDriver);
+  }
+
+  /**
+   * remove access to a media stream and stop all running video tracks
+   */
+  async stop(): Promise<void> {
+    if (this.inputMediaStream) {
+      for (const track of this.inputMediaStream.getVideoTracks()) {
+        track.stop();
+      }
+    }
+    this.inputMediaStream = null;
```

Comment:
So here is a big question:, when you stop, you set `this.inputMediaStream = null;`. Then most of them time (I am not sure if all); when you call `this.premiumVideoTranformDevice.stop()` you immidiately set `this.premiumVideoTransformDevice = null`. Can you just set `this = null` inside of this `stop` function?